packages/core/src/indexing-store/historical.ts

```diff
@@ -753,133 +793,169 @@ export const createHistoricalIndexingStore = ({
-                      .catch((_error) => {
-                        const error = _error as Error;
-                        common.logger.error({
-                          service: "indexing",
-                          msg: "Internal error occurred while flushing cache. Please report this error here: https://github.com/ponder-sh/ponder/issues",
-                        });
-                        throw new FlushError(error.message);
-                      }),
-                ),
-              );
-            }
+            promises.push(
+              database.qb.user.wrap(
+                { method: `${tableNameCache.get(table)}.flush()` },
+                async () => {
+                  const client = await (
+                    database.driver as { internal: Pool }
+                  ).internal.connect();
+
+                  const copyStream = client.query(
+                    copy.from(
+                      `COPY "${getTableConfig(table).schema ?? "public"}"."${getTableName(table)}" FROM STDIN WITH (FORMAT csv)`,
+                    ),
+                  );
+
+                  try {
+                    await pipeline(insertStream, copyStream);
+                  } catch (_error) {
+                    const error = _error as Error;
+                    common.logger.error({
+                      service: "indexing",
+                      msg: "Internal error occurred while flushing cache. Please report this error here: https://github.com/ponder-sh/ponder/issues",
+                    });
+                    throw new FlushError(error.message);
+                  } finally {
+                    client.release();
+                  }
+                },
+              ),
+            );
           }
 
-          if (updateValues.length > 0) {
+          if (updateSize > 0) {
             common.logger.debug({
               service: "indexing",
-              msg: `Updating ${updateValues.length} cached '${tableNameCache.get(table)}' records in the database`,
+              msg: `Updating ${updateSize} cached '${tableNameCache.get(table)}' rows in the database`,
             });
 
-            const primaryKeys = primaryKeysCache.get(table)!;
-
-            const set: { [column: string]: SQL } = {};
-            for (const [columnName, column] of Object.entries(
-              getTableColumns(table),
-            )) {
-              set[columnName] = sql.raw(
-                `excluded."${getColumnCasing(column, "snake_case")}"`,
-              );
-            }
+            promises.push(
+              database.qb.user.wrap(
+                { method: `${tableNameCache.get(table)}.flush()` },
+                async () => {
+                  const client = await (
+                    database.driver as { internal: Pool }
+                  ).internal.connect();
+
+                  try {
+                    await client.query(`
+CREATE TEMP TABLE "${tableNameCache.get(table)}" AS
```

Comment:
Is there any chance of a name collision with the temp table and an actual table?

Reply:
No, the temp tables use a separate namespace that takes precedence over others